### PR TITLE
Add new CoveredNetworks option

### PR DESCRIPTION
**Pull request by readams**
* Search by CIDR rather than just by IP.

Signed-off-by: Rob Adams <readams@readams.net>

### CODE REVIEW

File: brute.go [hunk 1]
```diff
@@ -87,6 +87,24 @@ func (b *bruteRanger) ContainingNetworks(ip net.IP) ([]RangerEntry, error) {
...
+// specified network.
+func (b *bruteRanger) CoveredNetworks(network net.IPNet) ([]RangerEntry, error) {
+	entries, err := b.getEntriesByVersion(network.IP)
+	if err != nil {
+		return nil, err
+	}
+	var results []RangerEntry
+	for _, entry := range entries {
+		entrynetwork := entry.Network()
+		if network.Contains(entrynetwork.IP) {
```
Comment (yl2chen) [inline brute.go]: This is incorrect, e.g. 192.168.0.0/15 is not covered by 192.168.0.0/16, but 192.168.0.0/16 does contain ip 192.168.0.0 (IP member of 192.168.0.0/15), you would need to compare the mask.  I would create a function on the Network struct `ContainsNetwork(Network)` to encapsulate the logic.

Adding the following test case to your test set would show this failure.
```
{
	rnet.IPv4,
	[]string{
		"192.168.0.0/15",
	},
	"192.168.0.0/16",
	[]string{},
	"should fail",
}
```

File: trie.go [hunk 2]
```diff
@@ -176,6 +184,31 @@ func (p *prefixTrie) containingNetworks(number rnet.NetworkNumber) ([]RangerEntr
 	return results, nil
 }
 
+func (p *prefixTrie) coveredNetworks(network rnet.Network) ([]RangerEntry, error) {
+	var results []RangerEntry
+	if p.hasEntry() && network.Contains(p.network.Number) {
```
Comment (yl2chen) [inline trie.go]: Same failure reason as above, would also show up with the added test case.  Should test by using the added correct ContainsNetwork function.

Also keep in mind that this is a trie, so if network represented at the current node is covered by the given network, all its children networks will also be covered by the given network, so any more inclusion testing is unnecessary.

File: brute.go (hunk 1 above)
Comment (yl2chen) [reply brute.go]: The following might not be optimized but should work
```
func (n Network) ContainsNetwork(n1 Network) bool {
	if len(n.Number) != len(n1.Number) {
		return false
	}
	nMaskSize, _ := n.IPNet.Mask.Size()
	n1MaskSize, _ := n1.IPNet.Mask.Size()
	if n1MaskSize < nMaskSize {
		return false
	}
	return n1.Masked(nMaskSize).Equal(n)
}
```

Comment (yl2chen) [reply brute.go]: Test case
```
func TestNetworkContainsNetwork(t *testing.T) {
	cases := []struct {
		n1           string
		n2           string
		n1ContainsN2 bool
	}{
		{"192.128.0.0/24", "192.128.0.0/24", true},
		{"192.128.0.0/24", "192.128.0.0/25", true},
		{"192.128.0.0/24", "192.128.0.0/23", false},
		{"192.128.0.0/24", "192.127.0.0/24", false},
		{"192.128.0.0/16", "192.128.199.0/24", true},
	}
	for _, tc := range cases {
		name := fmt.Sprintf("%s contains? %s", tc.n1, tc.n2)
		t.Run(name, func(t *testing.T) {
			_, ipNet1, _ := net.ParseCIDR(tc.n1)
			_, ipNet2, _ := net.ParseCIDR(tc.n2)
			assert.Equal(t, tc.n1ContainsN2, NewNetwork(*ipNet1).ContainsNetwork(NewNetwork(*ipNet2)))
		})
	}
}
```

File: net/ip.go [hunk 3]
```diff
@@ -208,6 +208,14 @@ func (n Network) Contains(nn NetworkNumber) bool {
 	return true
 }
 
+// Contains returns true if NetworkNumber is in range of Network, false
+// otherwise.
+func (n Network) Covers(o Network) bool {
```
Comment (yl2chen) [inline net/ip.go]: Add length check to make sure comparing IPv4 to IPv6 will return false, the reason being IPv4 and IPv6 example with same prefix could pass this test, but they don't cover each other
```
if len(n.Number) != len(n1.Number) {
	return false
}
```

File: net/ip_test.go [hunk 4]
```diff
@@ -412,6 +412,31 @@ func TestPreviousIP(t *testing.T) {
...
 }
 
+func TestNetworkCovers(t *testing.T) {
+	cases := []struct {
+		network string
+		covers  string
+		result  bool
+		name    string
+	}{
+		{"10.0.0.0/24", "10.1.0.0", true, "contains"},
```
Comment (yl2chen) [inline net/ip_test.go]: How is this test case passing?

**Comment by readams** [comment]
Thanks for catching that issue.  I just saw your ContainsNetwork which looks similar to what I added though I think the correct behavior should be the <= check so that a network covers itself.  A proper cover might be an interesting idea but probably not the most generally useful one.

File: net/ip_test.go (hunk 4 above)
Comment (readams) [reply net/ip_test.go]: I accidentally pushed the wrong patch to my tree; wasn't expecting it to get picked up :-).  The correct patch is there now.

**Comment by readams** [comment]
I'm pushing the changes as separate commits so it's easier to review but happy to squash them if you prefer or you can when the merge happens.

**Comment by yl2chen** [comment]
Hi Rob, I have added couple commits to your changes, and created another pull request https://github.com/yl2chen/cidranger/pull/9

**Pull request merged**